content/actions/guides/building-and-testing-powershell.md

```diff
@@ -0,0 +1,213 @@
+
+|| Ubuntu | Mac | Windows |
+|------|-------|------|----------|
+|**PowerShell system modules** |`/opt/microsoft/powershell/7/Modules/*`|`/usr/local/microsoft/powershell/7/Modules/*`|`C:\program files\powershell\7\Modules\*`|
+|**PowerShell add-on modules**|`/usr/local/share/powershell/Modules/*`|`/usr/local/share/powershell/Modules/*`|`C:\Modules\*`|
+|**User-installed modules**|`/home/runner/.local/share/powershell/Modules/*`|`/Users/runner/.local/share/powershell/Modules/*`|`C:\Users\runneradmin\Documents\PowerShell\Modules\*`|
+
+### Installing dependencies
+
+{% data variables.product.prodname_dotcom %}-hosted runners have PowerShell 7 and Pester installed. You can use `Install-Module` to install additional dependencies from the PowerShell Gallery before building and testing your code. For example, the YAML below installs the `SqlServer` and `PSScriptAnalyzer` modules.
+
+You can also cache dependencies to speed up your workflow. For more information, see "[Caching dependencies to speed up your workflow](/actions/automating-your-workflow-with-github-actions/caching-dependencies-to-speed-up-workflows)."
+
+{% raw %}
+```yaml
+steps:
+- uses: actions/checkout@v2
+- name: Install dependencies
+  shell: pwsh
+  run: |
+    Set-PSRepository PSGallery -InstallationPolicy Trusted
+    Install-Module SqlServer, PSScriptAnalyzer
+```
+{% endraw %}
+
+{% note %}
+
+**Note:** By default, no repositories are trusted by PowerShell. When installing modules from the PowerShell Gallery, you must explicitly set the installation policy for `PSGallery` to `Trusted`.
+
+{% endnote %}
+
+#### Caching Dependencies
+
+You can cache PowerShell module dependencies using a unique key, and restore the dependencies when you run future workflows using the [`cache`](https://github.com/marketplace/actions/cache) action. For more information, see "[Caching dependencies to speed up workflows](/actions/automating-your-workflow-with-github-actions/caching-dependencies-to-speed-up-workflows)."
+
+PowerShell caches dependencies in different locations, depending on the operating system of the runner. The path you'll need to cache may differ from the Ubuntu example below depending on the operating system you use. For more information, see [PowerShell caching examples](https://github.com/actions/cache/blob/main/examples.md#PowerShell).
+
+{% raw %}
+```yaml
+steps:
+  - uses: actions/checkout@v2
+  - name: Setup PowerShell module cache
+    id: cacher
+    uses: actions/cache@v2
+    with:
+      path: "~/.local/share/powershell/Modules"
+      key: ${{ runner.os }}-SqlServer-PSScriptAnalyzer
+  - name: Install required PowerShell modules
+    if: steps.cacher.outputs.cache-hit != 'true'
+    shell: pwsh
+    run: |
+      Set-PSRepository PSGallery -InstallationPolicy Trusted
+      Install-Module SqlServer, PSScriptAnalyzer -ErrorAction Stop
+```
+{% endraw %}
+
+### Testing your code
+
+You can use the same commands that you use locally to build and test your code.
```

Comment:
What do you think about explaining this code in this section:

```
- name: Perform a Pester test from the command-line to ensure expected results
      shell: pwsh
      run: Get-ChildItem | Select-Object -ExpandProperty Name -First 1 | Should -Be 'bin'
    - name: Perform advanced tests
      shell: pwsh
      run: |
        Invoke-Pester Unit.Tests.ps1 -Passthru
        Invoke-Pester Integration.Tests.ps1 -Passthru
```

In the other guides, we use this section to talk about the test steps and break down any commands that might need more description.

Reply:
That top one is such a simplified example, I'm not sure it's worth explaining - nobody will ever run this test. It is, however, the right structure for a test and is intended to give people ideas. 

Perhaps we remove the first test (that is unlikely to be used anyway) and just focus on the Unit and Integration tests. Those tests I wouldn't mind explaining. Do I do it in the comments?

Reply:
I think the code explanations for this example could be in a bulleted list that follows after the code block, and would only need to be for the PowerShell-specific `run:` commands. For example: (assuming we use a `test-path` example as suggested in another comment)

- `run: Test-Path resultsfile.log | Should -Be $true` - Checks whether the file named `resultsfile.log` is present.
- `Invoke-Pester Unit.Tests.ps1 -Passthru` - Runs the tests defined in the `Unit.Tests.ps1` file.

Reply:
Please add this as well, if you can., I'm not sure I'm following exactly, @martin389

Reply:
👍  Added the list of explanations, and an example screenshot and `Unit.Tests.ps1` entry.